### ISSUE TITLE
Submission：geospark

### ISSUE THREAD

**Issue by harryprince** [ropensci/software-review#288]
Submitting Author: Harry Zhu (@harryprince )  
Repository:   [geospark](https://github.com/harryprince/geospark)
Version submitted:   
Editor: TBD  
Reviewer 1: TBD  
Reviewer 2: TBD  
Archive: TBD  
Version accepted: TBD   

---

-   Paste the full DESCRIPTION file inside a code block below:

```
Package: geospark
Type: Package
Title: Bring Local Sf to Spark
Version: 0.0.2
Author: Harry Zhu
Maintainer: Harry Zhu <7harryprince@gmail.com>
Description: This package aims at bringing local sf to distributed spark cluster for easier geocomputing.
License: Apache License
Encoding: UTF-8
LazyData: true
Depends:
  R (>= 3.1.2)
Imports:
  sparklyr >= 1.0.0
```

## Scope 

- Please indicate which category or categories from our [package fit policies](https://ropensci.github.io/dev_guide/policies.html#package-categories) this package falls under: (Please check an appropriate box below. If you are unsure, we suggest you make a pre-submission inquiry.):

	- [ ] data retrieval
	- [ ] data extraction
	- [ ] database access
	- [x] data munging
	- [ ] data deposition
	- [ ] reproducibility
	- [x] geospatial data
	- [ ] text analysis
	

- Explain how and why the package falls under these categories (briefly, 1-2 sentences):

It brings distributed GIS database to the geospatial scientist who is known PostGIS or sf R package.

1. data munging based on sparklyr package
2.  geospatial data munging

the relationship between sparklyr and geospark is platform and plugin. [GeoSpark](https://datasystemslab.github.io/GeoSpark/) is a scala package developed by [Dr. Yu](https://github.com/jiayuasu), and I found it is very convenient for the geospatial scientist who already known PostGIS or sf R package.

-   Who is the target audience and what are scientific applications of this package?  

the geospatial scientist or transportation engineer.

The reason that I start to develop this package is making spatial operation with index easier. More generally, it is driven by bike-sharing parking business which needs to count how much bikes are within some polygons under a high presion level.

-   Are there other R packages that accomplish the same thing? If so, how does yours differ or meet [our criteria for best-in-category](https://ropensci.github.io/dev_guide/policies.html#overlap)?

1. Better in performance：[sf](https://github.com/r-spatial/sf) and [RPostGIS](https://github.com/mablab/rpostgis)  are doing a similar job on geospatial data processing scope too, however, which is not design for big data and only running on a single machine. In that, geospark R package is a good alternative for that.
2. Broader in functionality：[sparksf](https://github.com/etiennebr/sparksf) is doing a similar thing, however, geospark package support more GIS functions than that, such as `st_isvalid` and so on.

**Better in performance**:
GeoSpark is a cluster computing system for processing large-scale spatial data with a **distributed spatial index**. Here is more detail example of performance benchmarks with PostGIS and ESRI UDF: https://github.com/harryprince/geospark#benchmarks 

**Broader in functionality**:
`geospark` R package now is officially added to geospark scala package under latest [GeoSpark 1.2.0](https://github.com/DataSystemsLab/GeoSpark/tree/1.2.0-spark-2.3), which support more GIS SQL functions like `st_isvalid` and so on, you can see more features in [release notes](https://github.com/harryprince/geospark/releases).

**Great Usability and novel approach**
Here is a good example, which support `dplyr` and `sf` style grammar on spark in [geospark R package doc](https://github.com/harryprince/geospark#the-tidyverse-mode):

```
library(dplyr)
polygons_wkt <- mutate(polygons_wkt, y = st_geomfromwkt(geom, "4326"))
points_wkt <- mutate(points_wkt, x = st_geomfromwkt(geom, "4326"))

sc_res = inner_join(polygons_wkt, points_wkt, by = sql("st_contains(y, x)")) %>%
  group_by(area, state) %>%
  summarise(cnt = n())
  
sc_res %>%
  head()
```

```
    # Source: spark<?> [?? x 3]
    # Groups: area
      area            state   cnt
      <chr>           <chr> <dbl>
    1 texas area      TX       10
    2 dakota area     SD        1
    3 dakota area     ND       10
    4 california area CA       10
    5 new york area   NY        9
```

The reason that `geospark` R package can support dplyr and sf style grammar is  `dbplyr` can push local R functions into  `sparklyr` and parsed by the GIS SQL engine.

-   If you made a pre-submission enquiry, please paste the link to the corresponding issue, forum post, or other discussion, or @tag the editor you contacted.

* https://github.com/rstudio/sparklyr/issues/1700
* https://github.com/rstudio/sparklyr/issues/1475

## Technical checks

Confirm each of the following by checking the box.  This package:

- [x] does not violate the Terms of Service of any service it interacts with. 
- [ ] has a CRAN and OSI accepted license.
- [x] contains a [README with instructions for installing the development version](https://ropensci.github.io/dev_guide/building.html#readme). 
- [x] includes [documentation with examples for all functions](https://ropensci.github.io/dev_guide/building.html#documentation).
- [ ] contains a vignette with examples of its essential functions and uses.
- [ ] has a [test suite](https://ropensci.github.io/dev_guide/building.html#testing).
- [ ] has [continuous integration](https://ropensci.github.io/dev_guide/ci.html), including reporting of test coverage using services such as Travis CI, Coveralls and/or CodeCov.

## Publication options

- [x] Do you intend for this package to go on CRAN?  
- [ ] Do you wish to automatically submit to the [Journal of Open Source Software](http://joss.theoj.org/)? If so:

<details>
 <summary>JOSS Options</summary>  

  - [ ] The package has an **obvious research application** according to [JOSS's definition](https://joss.readthedocs.io/en/latest/submitting.html#submission-requirements).
    - [ ] The package contains a `paper.md` matching [JOSS's requirements](https://joss.readthedocs.io/en/latest/submitting.html#what-should-my-paper-contain) with a high-level description in the package root or in `inst/`.
    - [ ] The package is deposited in a long-term repository with the DOI: 
    - (*Do not submit your package separately to JOSS*)  
  
</details>

- [ ] Do you wish to submit an Applications Article about your package to [Methods in Ecology and Evolution](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/)? If so:  

<details>
<summary>MEE Options</summary>  

- [ ] The package is novel and will be of interest to the broad readership of the journal. 
- [ ] The manuscript describing the package is no longer than 3000 words.
- [ ] You intend to archive the code for the package in a long-term repository which meets the requirements of the journal (see [MEE's Policy on Publishing Code](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/journal-resources/policy-on-publishing-code.html))
- (*Scope: Do consider MEE's [Aims and Scope](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/aims-and-scope/read-full-aims-and-scope.html) for your manuscript. We make no guarantee that your manuscript will be within MEE scope.*)
- (*Although not required, we strongly recommend having a full manuscript prepared when you submit here.*)
- (*Please do not submit your package separately to Methods in Ecology and Evolution*)

</details>

## Code of conduct

- [x] I agree to abide by [rOpenSci's Code of Conduct](https://ropensci.github.io/dev_guide/policies.html#code-of-conduct) during the review process and in maintaining my package should it be accepted.

**Comment by melvidoni** [comment]
Thank you for your submission @harryprince.

Before we can progress this further, we need for you to add more details on your package. Please, expand on the Description,  and answer the questions under the `Scope` title in your opening post for the issue.

**Comment by harryprince** [comment]
@melvidoni thanks a lot.

I modified the related question under the `Scope` title.

**Comment by melvidoni** [comment]
Thank you @harryprince for answering the questions under Scope. However, the editorial team still feels that more detail needs to be added. Consider the following questions when updating the Scope again:

- Why are you stating that target audience?
- What does sparksf compared to geospark? Why is geospark better? Adding a short example would be ideal.
- How are those issues in sparklyr related to geospark?
- Please, extend the "Description" field inside the DESCRIPTION file.
- Why does geospark imports sparklyr?

Please, tag me again when you have addressed this.

**Comment by melvidoni** [comment]
@harryprince

Your response regarding the difference between `geospark` and `sparksf` is a critique of the package author, not a comparison of the two packages. Please refrain from any such personal attacks and provide specific examples of how the packages differ based on our overlap policies (https://ropensci.github.io/dev_guide/policies.html#overlap)


On another note, with the current documentation, `geospark` appears to be a very thin R package around what are essentially Java/scala programs, but there's basically no R code to review, no R-based documentation, and no R-specific API. So far, it is the Editors' interpretation that this is essentially a port of `sf` functions into scala. This is good, but possibly not in-scope for us.

Can you please refer to this view and provide a detailed justification on why it would be in-scope? This is for Editors to make an informed decision.

**Comment by harryprince** [comment]
@melvidoni thanks for your advice, it's my first time to submit R package, I am so sorry for what I have done. I will talk with sparksf author more.

**Comment by melvidoni** [comment]
@harryprince

Do not worry, it is something that you can solve. However, please tackle the following to-do list:

- [ ] Please, remove the "peer-reviewed" badge from your repository. This should _not_ be added until your package has been reviewed and approved by rOpenSci. We are still in a pre-processing stage.

- [ ] Your answer to the question _"Are there other R packages that accomplish the same thing? If so, how does yours differ or meet our criteria for best-in-category?_" is incomplete. Please, fill it with an answer that details the following answers:

  - What means "doing it better"?
  - What tasks are improved? Why?
  - What novel or new approach did you take?

- [ ] Please, explain with detail wheather your package is just a wrap of scala functions, or has extensive R code. Do this on the _"Scope"_ section.

**Comment by melvidoni** [comment]
I'm applying the `holding` badge for now.  If you'd like to resume the review process, we still need answers to the questions above. Also note your package will need to meet all the requirements under 'Technical Checks' to be reviewed

**Comment by harryprince** [comment]
@melvidoni I added a more detail example about how `geospark` works with `dplyr` and `sf` style grammar.

**Comment by melvidoni** [comment]
Thanks for your submission @harryprince, and for answering all of the editors' questions. Unfortunately, `geospark`is out-of-scope for rOpenSci software-review: there is a problem of thinness, as the only R code in the package a few lines of bindings, with the rest is the scala code for the spark plugin.  The examples show how that allows interactions with other R packages.

 Thanks again for submitting `geospark` to software-review! We hope you find another venue for it.
